### DISCUSSION TITLE
Groups of fields in a Singleton

### DISCUSSION THREAD

**Discussion by NeeshSamsi** [Thinkmill/keystatic#339] [Q&A] [answered]
I have a singleton for each page of a site. I would like to organize my fields into groups by the section of the page.

For example:
<img width="504" alt="image" src="https://github.com/Thinkmill/keystatic/assets/72866286/3268c30b-4eb3-458c-b82d-0006809da0f6">
I would like the first 4 fields under a Main/Hero sub-heading & then the next couple of fields under a Story subheading.

Is there a way to maybe have subheadings / sub-labels within a singleton for sections? Or something similar to achieve this?

**Comment by hauptrolle** [comment]
That would also be helpful for things like SEO data, where you want to group title, description, keywords etc..

**Comment by simonswiss** [comment]
Hey!

Yep, I have also personally suggested this and think this is very much needed. I am not sure _when_ we'll have something, but this is something the team has discussed implementing, it makes a lot of sense 👍

**Comment by JedWatson** [comment, marked as answer]
We're designing a solution for this at the moment.

@NeeshSamsi @hauptrolle would love some feedback if you have any thoughts, whether this will match what you're looking for:

Plan is to do something (syntax very draft) that lets you specify a group with a `label` and (optional) `description`, surrounding fields in the schema, like:

```
fields: {
  title: fields.text(),
  ...group({
    label: 'SEO',
    description: '...',
    fields: {
      description: fields.text(),
      // ...
    },
  },
}
```

And two mockups we're playing with:

![image](https://github.com/Thinkmill/keystatic/assets/872310/35b1dd3d-4289-4557-b830-7babedcc5c06)
![image (1)](https://github.com/Thinkmill/keystatic/assets/872310/7842e1fb-fc92-4978-b1ad-f828fa220459)

We're currently leaning towards just using whitespace, and trying to keep it really clean, not overwhelm the form with lines. But let us know what you think!

(also in those designs we're mocking up an update to the top part of the UI, and ways to show fields side by side, which... I guess let us know if you like that as well, but specifically looking for feedback on the field grouping here)

**edit** to be clear, this will actually work the same for entries in a collection *and* singletons 🙂

**Reply by MrMacStripe** [reply]
From my point of view neither whitespace or borders do the trick, everything is very white-on-white-ish.

What do you think about a different approach. I have two ideas in mind:
a) using a feed-style border on the left with bold markers next to the section headline/description (kind of like on gihubs landing page or https://tailwindcss.com/blog). This would provide some visual guidline where you are inside the page without created the everything-is-a-box effect.
b) Push the headline and the description to the left, creating an extra column next to the fields.

If you want am happy to visualize this. Is the mockup available somewhere to edit?

**Reply by JedWatson** [reply]
Thanks for the feedback @MrMacStripe

We don't have editable mockups I can share (those are exported from our Figma, and it'd be nice to open source that, but also quite a bit of work to clean up) but if you like, feel free to take those screenshots and edit them; move things around, etc.

Something we also have to figure out with all our designs is how they're responsive, from wide screen down through smaller / tablet sizes to mobile, so if your ideas include those considerations it would help too.

**Reply by MrMacStripe** [reply]
Sorry that it took so long, but please find my idea a) attached. It would need refinement for sure, but for me the groups are more obvious. I explored layout b) as well but did not find a good solution. 


Whitespace only would work as well probably, but the headings and the distance between groups would need to be bigger in my opinion.

![idea](https://github.com/Thinkmill/keystatic/assets/12595640/23edc8f0-ca1c-445b-831c-5d07368c6e78)

**Comment by hauptrolle** [comment]
The whitespace version looks very clean! Looking forward to the release 👍 
Displaying fields side by side is also a great change!

**Reply by simonswiss** [reply]
I also vote for the whitespace version — super clean 🤗

**Comment by NeeshSamsi** [comment]
Looking awesome! The whitespace version looks clean

**Comment by florian-lefebvre** [comment]
I suppose this will also be available for collections?
